lib/mix/lib/mix/local.ex

```diff
@@ -3,14 +3,43 @@ defmodule Mix.Local do
 
   @public_keys_html "https://s3.amazonaws.com/s3.hex.pm/installs/public_keys.html"
 
+  @type item :: :archive | :escript
+
   @doc """
-  The path for local archives.
+  Returns the name for an archive or an escript, based on on the project config.
+
+  ## Examples
+
+      iex> Mix.Local.name_for(:archive, [app: "foo", version: "0.1.0"])
+      "foo-0.1.0.ez"
+
+      iex> Mix.Local.name_for(:escript, [escript: [name: "foo"]])
+      "foo"
+
+  """
+  @spec name_for(item, Keyword.t) :: String.t
+  def name_for(:archive, project) do
+    version = if version = project[:version], do: "-#{version}"
+    "#{project[:app]}#{version}.ez"
+  end
+
+  def name_for(:escript, project) do
+    case get_in(project, [:escript, :name]) do
+      nil -> project[:app]
+      name -> name
+    end |> to_string()
+  end
 
-  Check `mix archive` for info.
+  @doc """
+  The path for local archives or escripts.
   """
-  def archives_path do
-    System.get_env("MIX_ARCHIVES") ||
-      Path.join(Mix.Utils.mix_home, "archives")
+  @spec path_for(item) :: String.t
+  def path_for(:archive) do
+    System.get_env("MIX_ARCHIVES") || Path.join(Mix.Utils.mix_home, "archives")
+  end
+
+  def path_for(:escript) do
+    Path.join(Mix.Utils.mix_home, "escripts")
```

Comment:
This has already been asked and answered [here](https://github.com/elixir-lang/elixir/pull/2974#issuecomment-68431565) (see the second quote).